README.md
# cartography

[![CRAN_Status_Badge](https://www.r-pkg.org/badges/version-ago/cartography)](https://cran.r-project.org/package=cartography)
[![Travis-CI Build Status](https://travis-ci.org/riatelab/cartography.svg?branch=master)](https://travis-ci.org/riatelab/cartography)
![](http://cranlogs.r-pkg.org/badges/cartography?color=brightgreen)
[![status](http://joss.theoj.org/papers/0c2d51fc23efb8e1f87d764da8414923/status.svg)](http://joss.theoj.org/papers/0c2d51fc23efb8e1f87d764da8414923)  
 

## *Create and integrate maps in your R workflow!*

[![Cartographic Mix](https://raw.githubusercontent.com/riatelab/cartography/master/img/map8.png "click on the map to see the code")](https://gist.github.com/rCarto/ef52aa4e96a7b628956fbf531143ae68)  

This package allows various **cartographic representations** such as 
proportional symbols, choropleth, typology, flows or discontinuities maps. It 
also offers several features enhancing the graphic presentation of maps like 
cartographic palettes, layout elements (scale, north arrow, title...), labels, 
legends or access to some cartographic APIs.

## Cheat Sheet
The [cheat sheet](http://riatelab.github.io/cartography/vignettes/cheatsheet/cartography_cheatsheet.pdf) displays a quick overview of `cartography`'s main features:
[![cartography cheat sheet](https://raw.githubusercontent.com/riatelab/cartography/master/img/cheat_sheet.png)](http://riatelab.github.io/cartography/vignettes/cheatsheet/cartography_cheatsheet.pdf)

## Vignette
The [vignette](https://CRAN.R-project.org/package=cartography/vignettes/cartography.html) 
contains commented scripts on how to build various types of maps with `cartography`:

<table>
<tbody>
<tr>
<td><img src="https://raw.githubusercontent.com/riatelab/cartography/master/img/map1.png" /></td>
<td><img src="https://raw.githubusercontent.com/riatelab/cartography/master/img/map2.png" /></td>
<td><img src="https://raw.githubusercontent.com/riatelab/cartography/master/img/map3.png" /></td>
</tr>
<tr>
<td><img src="https://raw.githubusercontent.com/riatelab/cartography/master/img/map4.png" /></td>
<td><img src="https://raw.githubusercontent.com/riatelab/cartography/master/img/map5.png" /></td>
<td><img src="https://raw.githubusercontent.com/riatelab/cartography/master/img/map6.png" /></td>
</tr>
</tbody>
</table>



## Blog Posts, Tutorials & Papers


[Giraud, T. and Lambert, N. (2017). “Reproducible Cartography.” In Peterson MP (ed.), _Advances in Cartography and GIScience. ICACI 2017. Lecture Notes in Geoinformation and Cartography._, pp.
173-183. doi: 10.1007/978-3-319-57336-6_13.](https://github.com/riatelab/ReproducibleCartography) (EN)      
[Cartographie reproductible](https://riatelab.github.io/cartographie-reproductible) (FR)  
[New version of the cartography package](https://rgeomatic.hypotheses.org/1205) (EN)    
[SF & cartography V2.0.0](https://rgeomatic.hypotheses.org/1149) (FR)   


## Demo
The following script creates a map of symbols that are proportional to values of a 
first variable and colored to reflect the discretization of a second variable.  


```r
library(cartography)
library(sf) 

# Data Import
# Import a shapefile // this one is distributed within the package
mtq <- st_read(system.file("shape/martinique.shp", package="cartography"))
# Compute the share of farmers in the active population
mtq$shareCS1 <- 100 * mtq$C13_CS1/mtq$C13_POP

########## Draft Map
# Plot the communes
plot(st_geometry(mtq))
# Plot symbols with choropleth coloration
propSymbolsChoroLayer(x = mtq, var = "C13_POP", var2 = "shareCS1")

# Add a layout
layoutLayer(title="Farmers in Martinique, 2013")
```

![](https://raw.githubusercontent.com/riatelab/cartography/master/img/map9.png)

```r
########## Final Map
# Set a custom color palette
cols <- carto.pal(pal1 = "wine.pal", n1 = 6)

# set plot margins
opar <- par(mar = c(0,0,1.2,0))

# Plot the communes
plot(st_geometry(mtq), col = "#5F799C", border = "white", 
     bg = "#A6CAE0", lwd = 0.5, add = FALSE)

# Plot symbols with choropleth coloration
propSymbolsChoroLayer(x = mtq, var = "C13_POP", var2 = "shareCS1", col = cols, 
                      inches = 0.4, method = "quantile", border = "grey50", 
                      lwd = 1, legend.var.pos = "topright", legend.var.style = "c",
                      legend.var2.pos = "left", legend.var2.title.txt =  
                        "Share of \nthe population\nworking in\nagriculture (%)", 
                      legend.var.title.txt = "Population aged\n15 and over") 

# Add a layout
layoutLayer(title="Farmers in Martinique, 2013", author = "cartography 2.1.3", 
            sources = "INSEE, 2016", scale = 5, tabtitle = TRUE, frame = FALSE)

# Add a north arrow
north(pos = "topleft")

# restore graphics parameters
par(opar)
```
![](https://raw.githubusercontent.com/riatelab/cartography/master/img/map7.png)


## Installation
* Development version on GitHub
```{r}
require(remotes)
install_github("riatelab/cartography")
```

* Stable version on [CRAN](https://CRAN.R-project.org/package=cartography/)
```{r}
install.packages("cartography")
```



## Alternatives Packages
* [tmap](https://github.com/mtennekes/tmap)    
* [ggplot2](https://github.com/tidyverse/ggplot2) + [ggspatial](https://github.com/paleolimbot/ggspatial)     
  


## Community Guidelines

One can contribute to the package through [pull requests](https://github.com/riatelab/cartography/pulls) and report issues or ask questions [here](https://github.com/riatelab/cartography/issues).






NEWS


cartography version 2.1..... (Release date: 2019)
==============
//- internaly, use sf instead of sp for hexagonal grid computation.



cartography version 2.1.3 (Release date: 2019-01-)
==============

Minor changes: 
- fix incorrect display of typo legends if legend.order is not provided.
- fix error in getLinkLayer when x and df ids do not perfectly match.
- only suggest OSM basemap sources without API key requirement.
- update README


cartography version 2.1.2 (Release date: 2018-09-17)
==============

Minor changes: 
- fix in getGridLayer for type ="hexagon" (sp import)
- addition of a cheat cheat reference in vignette (new vignette with a single link)
- Allow unprojected sf as input for getPencilLayer and prevent sampling error with a warning. 
- vignette correction (deprecated argument)



cartography version 2.1.1 (Release date: 2018-06-25)
==============

Minor changes:

- rgeos is back to Imports (was actually needed for some sp <> sf transformation)
- Add getPencilLayer function (color pencil layers)
- small fixes to follow sf API changes (warnings with st_centroids)



cartography version 2.1.0 (Release date: 2018-04-29)
==============

Major changes:

- sp and sf goes from Depends to Import, this implies the need to use library(sf) or library(sp) before using most of cartography functions. 
- rgeos is not used anymore
- packages sp, sf, graphics, stats, classInt and graphics are no more fully imported in the namespace

Minor changes:

- a lot of typos and mis-spelled words have been corrected
- propSymbols*Layer legends use the lwd argument
- LegendProp* have gained a border and an lwd argument




cartography version 2.0.2 (Release date: 2017-11-13)
==============

Minor Changes:

- bug fix in labelLayer() when x = SpatialPointsDataFrame.
- bug fix in propSymbols*Layer() when using a POINT layer with add=FALSE.
- bug fix in *Link*Layer() (df).
- add R, sf and sp version requirement + sf SystemRequirements
- allow dataframe AND tibbles as input for df arguùment in relevant functions
- add postitle to set the position of the title in laoutLayer()
- add tabtitle to set the size of the title banner in layoutLayer()



cartography version 2.0.1 (Release date: 2017-09-25)
==============

Minor Changes:

- bug fix for proportionnal symbols with negative values. 


cartography version 2.0.0 (Release date: 2017-09-19)
==============

Up to version 1.4.2 cartography was mainly based on sp and and rgeos for its 
spatial data management and geoprocessing operations. These dependencies 
are as much as possible replaced by sf functions since version 2.0.0.
Most functions are kept unchanged except for the addition of an x argument 
used to take sf objects as inputs.


Breaking changes:

* getBorders, getOuterBorders: an x argument is added in first position, x is added to take sf objects as inputs, if x is used then spdf is not used. spdfid is replaced by id. An sf object (MULTILINESTRING) of borders is returned.

* discLayer: spdf, spdfid1, spdfid2 and dfid are replaced by x (an sf object as outputed by getBorders). 

* getGridLayer: getGridLayer replaces getGridLayer + getGridData, x argument replaces spdf and can take an sf object or an sp object. var is added to enter the name(s) of numeric field(s) to compute into the grid. cellsize refers to the targeted area of the cells. The function returns an sf object

* getLinkLayer: spdf, spdf2, df, spdfid, spdf2id, dfids and dfide arguments are replaced by x, xid, df and dfid. x can be either a Spatial*DataFrame or an sf object. The function returns an sf LINESTRING instead of a SpatialLinesDataFrame.

* propLinkLayer, gradLinkLayer, gradLinkTypoLayer: spdf, df, spdfid, spdfids, spdfide, dfid, dfids and dfide arguments are replaced by x, df, xid and dfid. x must be an sf object (as outputed by getLinkLayer).

* legend* functions have been enhanced to be more customizable. It is now possible to lay legend on specified x & y. 

* propSymbolsLayer: breakval and col2 arguments are defunct, use propSymbolsTypoLayer if needed. 

Major changes: 

* choroLayer, typoLayer, propSymbolsLayer, propSymbolsChoroLayer, propSymbolsTypoLayer, dotDensityLayer, getTiles, getFigDim, labelLayer, smoothLayer: an x argument is added in first position of all these functions. x is added to take sf objects as inputs, if x is used then spdf, df, spdfid, id are not used. 

* labelLayer: possibility to add a colored halo around the labels, option to have non-overlapping labels.

* layoutLayer: the extent argument can take an sf object.

* Each function has an example based on sf object. 

* possibility to use text labels in legendChoro






cartography version 1.4.2 (Release date: 2017-03-29)
==============

Minor changes:

* Addition of 2 discretization methods (thanks to @MBunel): 'em' and 'arith'. The "arith" method is based on a arithmetic progression along the variable values. The "em" method is based on subset defined by the mean of a numeric vector.
* Update the list of the available tiles servers (getTiles). This list is reduced and could be extended again depanding on the rosm package evolution.



cartography version 1.4.1 (Release date: 2016-12-15)
==============

Major changes:  

* Addition of the hexagonal regular grid in getGridLayer.  
* Addition of barscale function to plot a custom scalebar.  
* Addition of north function to plot a custom north (or south) arrow.  

Minor changes:  

* small fix on the msd method (breaks order) in getBreaks function.  
* more precise sources for datasets
* renaming twincities data frame to twincities.df
* small fix for ids in getGridLayer


cartography version 1.4.0 (Release date: 2016-08-25)
==============

Major changes:  

* Introduce getOuterBorders that allows to build borders (as getBorders) between non-contiguous polygons.
* Addition of the theme parameter in layoutLayer to set col and coltitle according to a cartographic palette.
* discretization function become getBreaks. 

R/aaaaaaa.R







# #' @title Legend for Choropleth Maps
# #' @description Plot legend for choropleth maps.
# #' @name legendChoro
# #' @param pos position of the legend, one of "topleft", "top", 
# #' "topright", "left", "right", "bottomleft", "bottom", "bottomright".
# #' @param title.txt title of the legend.
# #' @param title.cex size of the legend title.
# #' @param values.cex size of the values in the legend.
# #' @param breaks break points in sorted order to indicate the intervals for assigning the colors. 
# #' Note that if there are nlevel colors (classes) there should be (nlevel+1) breakpoints.
# #' @param col a vector of colors. 
# #' @param cex size of the legend. 2 means two times bigger.
# #' @param values.rnd number of decimal places of the values in 
# #' the legend.
# #' @param nodata if TRUE a "no data" box or line is plotted.
# #' @param nodata.txt label for "no data" values. 
# #' @param nodata.col color of "no data" values.
# #' @param frame whether to add a frame to the legend (TRUE) or 
# #' not (FALSE).
# #' @param symbol type of symbol in the legend 'line' or 'box'
# #' @export
# #' @examples
# #' data("nuts2006")
# #' plot(nuts0.spdf, col = "grey")
# #' box()
# #' legendChoro(pos = "bottomleft", title.txt = "Title of the legend", title.cex = 0.8,
# #'             values.cex = 0.6, breaks = c(1,2,3,4,10.27,15.2),
# #'             col = carto.pal(pal1 = "orange.pal",n1 = 5), values.rnd =2,
# #'             nodata = TRUE, nodata.txt = "No data available", frame = TRUE, symbol="box")
# #' legendChoro(pos = "bottomright", title.txt = "Title of the legend", title.cex = 0.8,
# #'             values.cex = 0.6, breaks = c(1,2,5,7,10,15.27),
# #'             col = carto.pal(pal1 = "wine.pal",n1 = 5), values.rnd = 0,
# #'             nodata = TRUE, nodata.txt = "NA",nodata.col = "black",
# #'             frame = TRUE, symbol="line")
# legendChoro <- function(pos = "topleft", 
#                         title.txt = "Title of the legend", 
#                         title.cex = 0.8,
#                         values.cex = 0.6, 
#                         breaks, 
#                         col, 
#                         cex = 1,
#                         values.rnd =2, 
#                         nodata = TRUE, 
#                         nodata.txt = "No data", 
#                         nodata.col = "white",
#                         frame=FALSE,symbol="box"){
#   
#   positions <- c("bottomleft", "topleft", "topright", "bottomright", "left", 
#                  "right", "top", "bottom", "middle")
#   cex = 1
#   if(pos %in% positions){
#     # extent
#     legref <- function(cex){
#       x1 <- par()$usr[1]
#       x2 <- par()$usr[2]
#       y1 <- par()$usr[3]
#       y2 <- par()$usr[4]
#       xextent <- x2 - x1
#       yextent <- y2 - y1
#       paramsize1 = 30 / cex
#       paramsize2 <- paramsize1 * 40 / 25
#       width <- xextent / paramsize1
#       height <- width /1.5
#       delta1 <- min(yextent / paramsize2, xextent / paramsize2) # Gros eccart entre les objets
#       delta2 <- delta1 / 2 # Petit eccart entre les objets
#       return(list(x1 = x1, x2 = x2, y1 = y1, y2 = y2, xextent = xextent, 
#                   yextent = yextent, 
#                   width = width, height = height, delta1 = delta1, 
#                   delta2 = delta2))
#     }
#     lp <- legref(cex = cex)
#     names(lp)
#     
#     breaks = c(0.2,10,100,125,10)
#     values.rnd = 2
#     nodata = FALSE
#     values.cex = 0.6
#     title.cex = 0.8
#     title.txt = "Check my extra extra long title"
#     nodata.col = "red"
#     col = c("red", "yellow", "blue", "green")
#     
#     # Taille du bloc de legende
#     breaks <- as.numeric(round(breaks, values.rnd))
#     if (nodata == FALSE){nodata.txt <- NULL}
#     longval <- max(strwidth(c(breaks, nodata.txt), cex = values.cex))
#     
#     
#     # legend_xsize <- max(lp$width + strwidth(longval, cex = values.cex), 
#     #                     strwidth(title.txt, cex = title.cex) - lp$delta2) - lp$delta2
#     legend_xsize <- max(lp$width + strwidth(longval, cex = values.cex) + lp$delta2, 
#                         strwidth(title.txt, cex = title.cex))  
#     
#     legend_ysize <- (length(breaks)-1) * lp$height +  strheight(title.txt, cex = title.cex)
#     
#     # legende_size augmente si un caisson no data
#     if (nodata == TRUE){legend_ysize <- legend_ysize + lp$height + lp$delta2 }
#     
#     # Position
#     if (pos == "bottomleft") {
#       xref <- lp$x1 + lp$delta1 
#       yref <- lp$y1 + lp$delta1
#     }
#     if (pos == "topleft") {
#       xref <- lp$x1 + lp$delta1
#       yref <- lp$y2 - 2 * lp$delta1 - legend_ysize
#     }
#     if (pos == "topright") {
#       xref <- x2 - 2*delta1 - legend_xsize
#       yref <- y2 -2*delta1 - legend_ysize
#     }
#     if (pos == "bottomright") {
#       xref <- x2 - 2*delta1 - legend_xsize 
#       yref <- y1 + delta1}
#     if (pos == "left") {
#       xref <- x1 + delta1 
#       yref <- (y1+y2)/2-legend_ysize/2 - delta2
#     }
#     if (pos == "right") {
#       xref <- x2 - 2*delta1 - legend_xsize
#       yref <- (y1+y2)/2-legend_ysize/2 - delta2
#     }
#     if (pos == "top") {
#       xref <- (x1+x2)/2 - legend_xsize/2 
#       yref <- y2 - 2*delta1 - legend_ysize
#     }
#     if (pos == "bottom") {
#       xref <- (x1+x2)/2 - legend_xsize/2 
#       yref <- y1 + delta1
#     }
#     if (pos == "middle") {
#       xref <- (x1+x2)/2 - legend_xsize/2 
#       yref <- (y1+y2)/2-legend_ysize/2 - delta2
#     }
#     
#     plot(1:10)
#     points(xref, yref)
#     
#     
#     # Frame
#     if (frame==TRUE){
#       rect(xref - lp$delta1, 
#            yref - lp$delta1, 
#            xref + legend_xsize + lp$delta1 * 2, 
#            yref + legend_ysize + lp$delta1 * 2, 
#            border = "black",  col="white")
#     }
#     
#     # Affichage du bloc de legende
#     
#     if (nodata == TRUE){
#       rect(xref,
#            yref,
#            xref + lp$width,
#            yref + lp$height,
#            col = nodata.col, border = "black", lwd = 0.4)
#       text(xref + width + delta2 ,yref + height/2 ,nodata.txt,adj=c(0,0.5),cex=values.cex)
#       yref <- yref + height + delta2
#     }
#     
#     if (symbol=="line"){
#       for (i in 0:(length(breaks)-2)){
#         segments(xref, yref + height/2+ i*height, xref + width, yref + i*height + height/2, lwd=5, col=col[i+1], lend=1)
#       }
#     } else { #box
#       for (i in 0:(length(breaks)-2)){
#         rect(xref,
#              yref + i * lp$height,
#              xref + lp$width,
#              yref + lp$height + i * lp$height,
#              col = col[i+1], border = "black", lwd = 0.4)
#       }
#       
#     }
#     
#     # Affichage des textes
#     for (i in 1:(length(breaks))){
#       text(x = xref + lp$width + lp$delta2 ,
#            y = yref + (i-1) * lp$height,
#            labels = breaks[i],
#            adj = c(0,0.5), cex = values.cex)
#     }
#     
#     # Affichage du titre
#     text(x = xref, y = yref + (length(breaks)-1) * lp$height + lp$delta1, 
#          labels = title.txt, adj = c(0,0), cex = title.cex)
#     
#   }
#   
# }